pkg/kv/dist_sender.go

```diff
@@ -1158,51 +1135,54 @@ func (ds *DistSender) sendToReplicas(
+	log.VEventf(ctx, 2, "r%d: sending batch %s to %s", rangeID, args.Summary(), transport.NextReplica())
 	transport.SendNext(ctx, done)
 
 	// Wait for completions. This loop will retry operations that fail
 	// with errors that reflect per-replica state and may succeed on
 	// other replicas.
-	sendNextTimer := timeutil.NewTimer()
 	slowTimer := timeutil.NewTimer()
-	defer sendNextTimer.Stop()
 	defer slowTimer.Stop()
 	slowTimer.Reset(base.SlowRequestThreshold)
 	for {
-		if timeout, ok := transport.SendNextTimeout(opts.SendNextTimeout); ok {
-			// Only start the send-next timer if we haven't exhausted the transport
-			// (i.e. there is another replica to send to).
-			sendNextTimer.Reset(timeout)
-		}
-
 		select {
-		case <-sendNextTimer.C:
-			sendNextTimer.Read = true
-			// On successive RPC timeouts, send to additional replicas if available.
-			if !transport.IsExhausted() {
-				ds.metrics.SendNextTimeoutCount.Inc(1)
-				log.VEventf(ctx, 2, "timeout, trying next peer: %s", transport.NextReplica())
-				pending++
-				transport.SendNext(ctx, done)
-			}
-
 		case <-slowTimer.C:
 			log.Warningf(ctx, "have been waiting %s sending RPC to r%d for batch: %s",
 				base.SlowRequestThreshold, rangeID, args)
 			ds.metrics.SlowRequestsCount.Inc(1)
 			defer ds.metrics.SlowRequestsCount.Dec(1)
 
 		case call := <-done:
-			pending--
-			err := call.Err
-			if err == nil {
+			if err := call.Err; err != nil {
+				// All connection errors except for an unavailable node (this
+				// is GRPC's fail-fast error), may mean that the request
+				// succeeded on the remote server, but we were unable to
+				// receive the reply. Set the ambiguous commit flag.
+				//
+				// We retry ambiguous commit batches to avoid returning the
+				// unrecoverable AmbiguousResultError. This is safe because
+				// repeating an already-successfully applied batch is
+				// guaranteed to return either a TransactionReplayError (in
+				// case the replay happens at the original leader), or a
+				// TransactionRetryError (in case the replay happens at a new
+				// leader). If the original attempt merely timed out or was
+				// lost, then the batch will succeed and we can be assured the
+				// commit was applied just once.
+				//
+				// The Unavailable code is used by GRPC to indicate that a
+				// request fails fast and is not sent, so we can be sure there
+				// is no ambiguity on these errors. Note that these are common
+				// if a node is down.
+				// See https://github.com/grpc/grpc-go/blob/52f6504dc290bd928a8139ba94e3ab32ed9a6273/call.go#L182
+				// See https://github.com/grpc/grpc-go/blob/52f6504dc290bd928a8139ba94e3ab32ed9a6273/stream.go#L158
+				if haveCommit && grpc.Code(err) != codes.Unavailable {
+					return nil, roachpb.NewAmbiguousResultError(fmt.Sprintf("error=%s", err))
+				}
+				log.ErrEventf(ctx, "RPC error: %s", err)
+			} else {
 				// Determine whether the error must be propagated immediately or whether
```

Comment:
This comment should probably be adjusted to reflect the fact there may not actually be an error. It's confusing as is.